include/swift/SILOptimizer/Utils/CanonicalizeInstruction.h

```diff
@@ -0,0 +1,87 @@
+//===-- CanonicalizeInstruction.h - canonical SIL peepholes -----*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// SSA-peephole transformations that yield a more canonical SIL representation.
+///
+/// Unlike simplifyInstruction, these transformations may effect any
+/// instruction, not only single-values, and may arbitrarily generate new SIL
+/// instructions.
+///
+/// Unlike SILCombine, these peepholes must work on 'raw' SIL form and should be
+/// limited to those necessary to aid in diagnostics and other mandatory
+/// pipelin/e passes. Optimization may only be done to the extent that it
+/// neither interferes with diagnostics nor increases compile time.
+///
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_SILOPTIMIZER_UTILS_CANONICALIZEINSTRUCTION_H
+#define SWIFT_SILOPTIMIZER_UTILS_CANONICALIZEINSTRUCTION_H
+
+#include "swift/SIL/SILBasicBlock.h"
+#include "swift/SIL/SILInstruction.h"
+#include "llvm/Support/Debug.h"
+
+namespace swift {
+
+/// Abstract base class. Implements all canonicalization transforms. Extended by
+/// passes to be notified of each SIL modification.
+struct CanonicalizeInstruction {
+  // May be overriden by passes.
+  static constexpr const char *defaultDebugType = "sil-canonicalize";
+  const char *debugType = defaultDebugType;
+
+  CanonicalizeInstruction(const char *passDebugType) {
+#ifndef NDEBUG
+    if (llvm::DebugFlag && !llvm::isCurrentDebugType(debugType))
+      debugType = passDebugType;
+#endif
+  }
+
+  virtual ~CanonicalizeInstruction();
+
+  /// Rewrite this instruction, based on its operands and uses, into a more
+  /// canonical representation.
+  ///
+  /// Return an iterator to the next instruction or to the end of the block.
+  /// The returned iterator will follow any newly added or to-be-deleted
+  /// instructions, regardless of whether the pass immediately deletes the
+  /// instructions or simply records them for later deletion.
+  ///
+  /// To (re)visit new instructions, override notifyNewInstruction().
+  ///
+  /// To determine if any transformation at all occurred, override
+  /// notifyNewInstruction(), killInstruction(), and notifyNewUsers().
+  ///
+  /// Warning: If the \p inst argument is killed and the client immediately
+  /// erases it, then it may be an invalid pointer upon return.
```

Comment:
Can you replace the 2nd it here with `\p inst`. Then it is really clear that inst may be an invalid pointer upon return.